plugin/conn_ip_example/conn_ip_example_test.go

```diff
@@ -11,46 +11,109 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package main_test
+package main
 
 import (
 	"context"
+	"fmt"
+	"strconv"
 	"testing"
 
 	. "github.com/pingcap/check"
 	"github.com/pingcap/tidb/plugin"
+	"github.com/pingcap/tidb/sessionctx/variable"
 )
 
-func LoadRunShutdownPluginExample() {
+type testConnIPExampleSuite struct{}
+
+var _ = SerialSuites(&testConnIPExampleSuite{})
+
+func (s *testConnIPExampleSuite) TestLoadPlugin(c *C) {
 	ctx := context.Background()
-	var pluginVarNames []string
+	pluginName := "conn_ip_example"
+	pluginVersion := uint16(1)
+	pluginSign := pluginName + "-" + strconv.Itoa(int(pluginVersion))
+
 	cfg := plugin.Config{
-		Plugins:        []string{"conn_ip_example-1"},
-		PluginDir:      "/home/robi/Code/go/src/github.com/pingcap/tidb/plugin/conn_ip_example",
-		PluginVarNames: &pluginVarNames,
+		Plugins:        []string{pluginSign},
+		PluginDir:      "",
+		PluginVarNames: &variable.PluginVarNames,
+		EnvVersion:     map[string]uint16{"go": 1112},
+	}
+
+	// setup load test hook.
+	loadOne := func(p *plugin.Plugin, dir string, pluginID plugin.ID) (manifest func() *plugin.Manifest, err error) {
+		return func() *plugin.Manifest {
+			m := &plugin.AuditManifest{
+				Manifest: plugin.Manifest{
+					Kind:       plugin.Audit,
+					Name:       pluginName,
+					Version:    pluginVersion,
+					SysVars:    map[string]*variable.SysVar{pluginName + "_key": {Scope: variable.ScopeGlobal, Name: pluginName + "_key", Value: "v1"}},
+					OnInit:     OnInit,
+					OnShutdown: OnShutdown,
+					Validate:   Validate,
+				},
+				OnGeneralEvent:    OnGeneralEvent,
+				OnConnectionEvent: OnConnectionEvent,
+			}
+			return plugin.ExportManifest(m)
+		}, nil
 	}
+	plugin.SetTestHook(loadOne)
 
+	// trigger load.
 	err := plugin.Load(ctx, cfg)
 	if err != nil {
-		panic(err)
+		fmt.Printf("load plugin [%s] fail\n", pluginSign)
+		fmt.Print(err)
 	}
 
-	// load and start TiDB domain.
-
 	err = plugin.Init(ctx, cfg)
 	if err != nil {
-		panic(err)
+		fmt.Printf("init plugin [%s] fail\n", pluginSign)
 	}
 
 	err = plugin.ForeachPlugin(plugin.Audit, func(auditPlugin *plugin.Plugin) error {
 		plugin.DeclareAuditManifest(auditPlugin.Manifest).OnGeneralEvent(context.Background(), nil, plugin.Log, "QUERY")
 		return nil
 	})
-	if err != nil {
-		panic(err)
+	connectionNum := 5
+	for i := 0; i < connectionNum; i++ {
+		err = plugin.ForeachPlugin(plugin.Audit, func(auditPlugin *plugin.Plugin) error {
+			err = plugin.DeclareAuditManifest(auditPlugin.Manifest).OnConnectionEvent(context.Background(), plugin.Connected, &variable.ConnectionInfo{Host: "localhost"})
+			if err != nil {
+				fmt.Printf("OnConnectionEvent error [%s]\n", err)
+			}
+			return nil
```

Comment:
return the error

Reply:
You should not fatal in the closure: propagate it to the outside(return it), so that the caller could process the error. @An-DJ